Sources/TuistDependencies/SwiftPackageManager/Utils/SwiftPackageManagerGraphGenerator.swift

```diff
@@ -72,239 +73,204 @@ public final class SwiftPackageManagerGraphGenerator: SwiftPackageManagerGraphGe
             packageInfo.info.products.forEach { result[$0.name] = packageInfo.name }
         }
 
-        let externalDependencies: [String: ExternalDependency] = Dictionary(uniqueKeysWithValues: try packageInfos.map { packageInfo in
-            let dependency = try Self.mapToExternalDependency(
+        return try packageInfos.reduce(DependenciesGraph.none) { result, packageInfo in
+            try Self.writeProject(
+                for: packageInfo.info,
                 name: packageInfo.name,
-                folder: packageInfo.folder,
-                packageInfo: packageInfo.info,
-                artifactsFolder: packageInfo.artifactsFolder,
+                at: packageInfo.folder,
                 productToPackage: productToPackage
             )
-            return (dependency.name, dependency)
-        })
-
-        return DependenciesGraph(externalDependencies: externalDependencies)
+            let packageDependenciesGraph = DependenciesGraph(
+                externalDependencies: packageInfo.info.products.reduce(into: [:]) { result, product in
+                    result[product.name] = product.targets.map { .project(target: $0, path: packageInfo.folder) }
+                }
+            )
+            return try result.merging(with: packageDependenciesGraph)
+        }
     }
 
-    // swiftlint:disable:next function_body_length
-    private static func mapToExternalDependency(
+    private static func writeProject(
+        for packageInfo: PackageInfo,
         name: String,
-        folder: AbsolutePath,
+        at folder: AbsolutePath,
+        productToPackage: [String: String]
+    ) throws {
+        let targets = try packageInfo.targets.compactMap { target in
+            try Self.targetDefinition(for: target, packageName: name, packageInfo: packageInfo, at: folder, productToPackage: productToPackage)
+        }
+        let project = ProjectDescription.Project(
+            name: name,
+            targets: targets,
+            resourceSynthesizers: []
+        )
+        let projectData = String(data: try JSONEncoder().encode(project), encoding: .utf8)!
```

Comment:
I believe you should be able to write data directly and conversion to a string is not needed ... or am I wrong?

Reply:
The `FileHandling` only exposes the `String`-based write, would you prefer to expand the protocol instead?